In [52]:
import pandas as pd

In [53]:
patients = pd.read_csv("data/patients.csv")
print(patients.columns)
patients

Index(['Id', 'BIRTHDATE', 'DEATHDATE', 'SSN', 'DRIVERS', 'PASSPORT', 'PREFIX',
       'FIRST', 'LAST', 'SUFFIX', 'MAIDEN', 'MARITAL', 'RACE', 'ETHNICITY',
       'GENDER', 'BIRTHPLACE', 'ADDRESS', 'CITY', 'STATE', 'COUNTY', 'FIPS',
       'ZIP', 'LAT', 'LON', 'HEALTHCARE_EXPENSES', 'HEALTHCARE_COVERAGE',
       'INCOME'],
      dtype='object')


Id   BIRTHDATE   DEATHDATE  \
0     bb30994a-ad91-d617-2466-e9ef4c3838dd  2010-10-24         NaN   
1     22f309fb-7b30-144b-cedb-ce932fe286e4  2009-03-16         NaN   
2     343787c0-ee04-90d8-61a5-4ed996438c63  1981-07-08         NaN   
3     162a2c8c-f58e-95e5-d952-13b08627fd1a  1978-03-07         NaN   
4     a1c82b81-cc4e-1eef-1046-a4081a4b12b4  1991-02-12         NaN   
...                                    ...         ...         ...   
5710  f4cee386-514c-de23-bf20-6d88f3153571  1986-07-19         NaN   
5711  7a9d48f0-623c-9dda-1de7-681f1d2e7f7b  1958-07-20  2015-04-09   
5712  27d948ff-e081-99c7-83a3-ef9e94a6b01d  2000-10-17         NaN   
5713  bd7ebd2a-6e12-2b09-a22f-a3ed83a93561  1958-07-20         NaN   
5714  30bfbb3e-2f6a-f599-c87e-3b3fdf0d0b33  1946-10-19         NaN   

              SSN    DRIVERS    PASSPORT PREFIX        FIRST           LAST  \
0     999-25-7461        NaN         NaN    NaN    Nathan164    Schuster709   
1     999-37-4271        NaN         NaN    NaN      Asia282       Lesch175   
2     999-84-8564  S99994136  X23175682X    Mr.      Abram53        Cole117   
3     999-54-8345  S99917811  X33252218X   Mrs.   Eufemia350  Jakubowski832   
4     999-71-9025  S99938018  X79840478X    Mr.     Javier97       Terán294   
...           ...        ...         ...    ...          ...            ...   
5710  999-22-2904  S99962544  X63396209X    Mr.   Octavio643       Muñoz828   
5711  999-71-2826  S99958264  X62286915X    Mr.       Abe604       Stark857   
5712  999-97-2074  S99956339  X38583829X    Mr.   Forrest301      Waters156   
5713  999-20-9898  S99913259  X87294055X    Mr.     Kelly223    Ondricka197   
5714  999-12-6776  S99914184  X77226361X   Mrs.  Graciela518       Zarate81   

     SUFFIX  ...           CITY       STATE                COUNTY    FIPS  \
0       NaN  ...         Walnut  California    Los Angeles County  6037.0   
1       NaN  ...        Soledad  California       Monterey County  6053.0   
2       NaN  ...        Hanford  California          Kings County  6031.0   
3       NaN  ...  San Francisco  California  San Francisco County  6075.0   
4       NaN  ...        Arcadia  California    Los Angeles County  6037.0   
...     ...  ...            ...         ...                   ...     ...   
5710    NaN  ...      San Diego  California      San Diego County  6073.0   
5711    NaN  ...     Costa Mesa  California         Orange County  6059.0   
5712    NaN  ...        Norwalk  California    Los Angeles County  6037.0   
5713    NaN  ...     Costa Mesa  California         Orange County  6059.0   
5714    NaN  ...         Carson  California    Los Angeles County  6037.0   

        ZIP        LAT         LON HEALTHCARE_EXPENSES HEALTHCARE_COVERAGE  \
0     91724  34.110204 -117.803515            25831.61             3234.86   
1     93960  36.342698 -121.434998            36690.25                0.00   
2     93230  36.260623 -119.615211             6289.69            54030.33   
3     94111  37.780116 -122.450481            16832.02          2339850.05   
4     91732  34.138577 -118.065934            66081.73            20372.03   
...     ...        ...         ...                 ...                 ...   
5710  92122  32.901176 -117.038117            60885.15            14885.42   
5711  92626  33.631114 -117.798555           152505.53          2443760.53   
5712  90703  33.836854 -118.078111            13482.06          1047683.40   
5713  92646  33.724060 -117.903008           354873.71          1065469.66   
5714  90248  33.868254 -118.249119           737028.50           851526.10   

      INCOME  
0     125285  
1      47629  
2      10371  
3      19571  
4     135002  
...      ...  
5710  894206  
5711  106186  
5712   87279  
5713  106186  
5714   88949  

[5715 rows x 27 columns]

In [54]:
conditions = pd.read_csv("data/conditions.csv")
print(conditions.columns)
#select patients with anxiety
conditions = conditions[conditions["DESCRIPTION"].str.contains("anxiety")]
conditions

Index(['START', 'STOP', 'PATIENT', 'ENCOUNTER', 'CODE', 'DESCRIPTION'], dtype='object')


START        STOP                               PATIENT  \
156     2017-04-22  2020-05-09  16c73553-9400-1f2c-bcf9-7b5ef75150d2   
844     2013-09-01  2019-10-06  586aef93-ba76-6c67-4d74-0d73019562fa   
1092    1985-07-22  1986-03-31  e400b4da-8aa0-e8a8-2f28-a427256fc3c8   
1676    2017-04-10  2018-09-03  b968acf5-d1e9-a555-aba3-4869bf93a908   
2154    2012-07-18  2014-07-30  fa14c8db-b4d2-e8c7-e23c-c0ee6b52234c   
...            ...         ...                                   ...   
279319  2014-02-18  2016-06-28  85f87a21-aedb-d02a-b384-a1502c8d7c60   
279454  2021-10-05  2021-10-19  85f87a21-aedb-d02a-b384-a1502c8d7c60   
279514  2014-12-13  2015-12-19  d7a4fbf8-3275-6a45-cae8-a9eab8c58721   
279832  2013-11-12  2014-11-18  27d948ff-e081-99c7-83a3-ef9e94a6b01d   
280037  2014-03-08  2014-06-28  30bfbb3e-2f6a-f599-c87e-3b3fdf0d0b33   

                                   ENCOUNTER      CODE  \
156     a2fc1abb-5b50-a3df-c49b-51daa9fa0690  80583007   
844     939f6784-2fa6-f321-bbf4-a6a760d59f34  80583007   
1092    456fb401-8cf7-21f4-c3b7-bd743c86ccb2  80583007   
1676    b34ffbe3-c1ab-5735-a3a5-5301b8b0a3c8  80583007   
2154    e96b0f04-9364-1e0a-94b5-bde1ac9e6b01  80583007   
...                                      ...       ...   
279319  ce783f75-9642-e134-391f-fe0f17b82608  80583007   
279454  7c84b245-8de1-61a1-29b5-5e49491625e6  80583007   
279514  36758771-34aa-a009-bf08-f43c4a057617  80583007   
279832  8bcfec76-e5bf-6321-d68c-be86f4a8f9a6  80583007   
280037  246c5e26-ae65-db8d-9b76-fd018a8c0be0  80583007   

                             DESCRIPTION  
156     Severe anxiety (panic) (finding)  
844     Severe anxiety (panic) (finding)  
1092    Severe anxiety (panic) (finding)  
1676    Severe anxiety (panic) (finding)  
2154    Severe anxiety (panic) (finding)  
...                                  ...  
279319  Severe anxiety (panic) (finding)  
279454  Severe anxiety (panic) (finding)  
279514  Severe anxiety (panic) (finding)  
279832  Severe anxiety (panic) (finding)  
280037  Severe anxiety (panic) (finding)  

[820 rows x 6 columns]

In [55]:
anxiety_patients = pd.merge(patients, conditions, left_on='Id', right_on='PATIENT')
#select specific columns
anxiety_patients = anxiety_patients[['Id', 'ZIP', 'RACE', 'GENDER', 'ETHNICITY', 'INCOME', 'HEALTHCARE_EXPENSES', 'HEALTHCARE_COVERAGE', 'DESCRIPTION']]
anxiety_patients

Id    ZIP   RACE GENDER    ETHNICITY  \
0    16c73553-9400-1f2c-bcf9-7b5ef75150d2  95616  white      F     hispanic   
1    586aef93-ba76-6c67-4d74-0d73019562fa  92336  asian      F     hispanic   
2    e400b4da-8aa0-e8a8-2f28-a427256fc3c8  92128  white      M  nonhispanic   
3    b968acf5-d1e9-a555-aba3-4869bf93a908  92025  white      M  nonhispanic   
4    fa14c8db-b4d2-e8c7-e23c-c0ee6b52234c  92692  white      F  nonhispanic   
..                                    ...    ...    ...    ...          ...   
815  85f87a21-aedb-d02a-b384-a1502c8d7c60  90001  asian      M  nonhispanic   
816  85f87a21-aedb-d02a-b384-a1502c8d7c60  90001  asian      M  nonhispanic   
817  d7a4fbf8-3275-6a45-cae8-a9eab8c58721  90803  white      F  nonhispanic   
818  27d948ff-e081-99c7-83a3-ef9e94a6b01d  90703  white      M     hispanic   
819  30bfbb3e-2f6a-f599-c87e-3b3fdf0d0b33  90248  white      F     hispanic   

     INCOME  HEALTHCARE_EXPENSES  HEALTHCARE_COVERAGE  \
0     45504            286834.48            130435.94   
1     61586            623248.61            992167.75   
2    162992            126001.52            314422.42   
3    162992            179541.06            311634.06   
4     66594             72988.27            333305.01   
..      ...                  ...                  ...   
815   34984            173668.03            513869.30   
816   34984            173668.03            513869.30   
817   82156            717729.34            984149.07   
818   87279             13482.06           1047683.40   
819   88949            737028.50            851526.10   

                          DESCRIPTION  
0    Severe anxiety (panic) (finding)  
1    Severe anxiety (panic) (finding)  
2    Severe anxiety (panic) (finding)  
3    Severe anxiety (panic) (finding)  
4    Severe anxiety (panic) (finding)  
..                                ...  
815  Severe anxiety (panic) (finding)  
816  Severe anxiety (panic) (finding)  
817  Severe anxiety (panic) (finding)  
818  Severe anxiety (panic) (finding)  
819  Severe anxiety (panic) (finding)  

[820 rows x 9 columns]

NOTE: Multiple observations for same patient can exist 

In [56]:
observations = pd.read_csv("data/observations.csv")
print(observations.columns)
#select observations recorded for anxiety
observations = observations[observations["DESCRIPTION"].str.contains("anxiety")]


Index(['DATE', 'PATIENT', 'ENCOUNTER', 'CATEGORY', 'CODE', 'DESCRIPTION',
       'VALUE', 'UNITS', 'TYPE'],
      dtype='object')


In [57]:
anxiety_observations = pd.merge(anxiety_patients, observations, left_on='Id', right_on='PATIENT')
#select columns of interest
anxiety_observations = anxiety_observations[['Id', 'ZIP', 'RACE', 'GENDER', 'ETHNICITY', 'INCOME', 'HEALTHCARE_EXPENSES', 'HEALTHCARE_COVERAGE', 'DESCRIPTION_x', 'DESCRIPTION_y', 'VALUE']]
anxiety_observations

Id    ZIP   RACE GENDER ETHNICITY  \
0      16c73553-9400-1f2c-bcf9-7b5ef75150d2  95616  white      F  hispanic   
1      16c73553-9400-1f2c-bcf9-7b5ef75150d2  95616  white      F  hispanic   
2      16c73553-9400-1f2c-bcf9-7b5ef75150d2  95616  white      F  hispanic   
3      16c73553-9400-1f2c-bcf9-7b5ef75150d2  95616  white      F  hispanic   
4      16c73553-9400-1f2c-bcf9-7b5ef75150d2  95616  white      F  hispanic   
...                                     ...    ...    ...    ...       ...   
14841  30bfbb3e-2f6a-f599-c87e-3b3fdf0d0b33  90248  white      F  hispanic   
14842  30bfbb3e-2f6a-f599-c87e-3b3fdf0d0b33  90248  white      F  hispanic   
14843  30bfbb3e-2f6a-f599-c87e-3b3fdf0d0b33  90248  white      F  hispanic   
14844  30bfbb3e-2f6a-f599-c87e-3b3fdf0d0b33  90248  white      F  hispanic   
14845  30bfbb3e-2f6a-f599-c87e-3b3fdf0d0b33  90248  white      F  hispanic   

       INCOME  HEALTHCARE_EXPENSES  HEALTHCARE_COVERAGE  \
0       45504            286834.48            130435.94   
1       45504            286834.48            130435.94   
2       45504            286834.48            130435.94   
3       45504            286834.48            130435.94   
4       45504            286834.48            130435.94   
...       ...                  ...                  ...   
14841   88949            737028.50            851526.10   
14842   88949            737028.50            851526.10   
14843   88949            737028.50            851526.10   
14844   88949            737028.50            851526.10   
14845   88949            737028.50            851526.10   

                          DESCRIPTION_x  \
0      Severe anxiety (panic) (finding)   
1      Severe anxiety (panic) (finding)   
2      Severe anxiety (panic) (finding)   
3      Severe anxiety (panic) (finding)   
4      Severe anxiety (panic) (finding)   
...                                 ...   
14841  Severe anxiety (panic) (finding)   
14842  Severe anxiety (panic) (finding)   
14843  Severe anxiety (panic) (finding)   
14844  Severe anxiety (panic) (finding)   
14845  Severe anxiety (panic) (finding)   

                                           DESCRIPTION_y VALUE  
0      Generalized anxiety disorder 7 item (GAD-7) to...   3.0  
1      Generalized anxiety disorder 7 item (GAD-7) to...   3.0  
2      Generalized anxiety disorder 7 item (GAD-7) to...  21.0  
3      Generalized anxiety disorder 7 item (GAD-7) to...   2.0  
4      Generalized anxiety disorder 7 item (GAD-7) to...   3.0  
...                                                  ...   ...  
14841  Generalized anxiety disorder 7 item (GAD-7) to...   1.0  
14842  Generalized anxiety disorder 7 item (GAD-7) to...   2.0  
14843  Generalized anxiety disorder 7 item (GAD-7) to...   3.0  
14844  Generalized anxiety disorder 7 item (GAD-7) to...   8.0  
14845  Generalized anxiety disorder 7 item (GAD-7) to...   0.0  

[14846 rows x 11 columns]

In [58]:
anxiety_observations = anxiety_observations.rename(columns={'DESCRIPTION_x':'CONDITION', 'DESCRIPTION_y':'OBSERVED_CONDITION'})
anxiety_observations

Id    ZIP   RACE GENDER ETHNICITY  \
0      16c73553-9400-1f2c-bcf9-7b5ef75150d2  95616  white      F  hispanic   
1      16c73553-9400-1f2c-bcf9-7b5ef75150d2  95616  white      F  hispanic   
2      16c73553-9400-1f2c-bcf9-7b5ef75150d2  95616  white      F  hispanic   
3      16c73553-9400-1f2c-bcf9-7b5ef75150d2  95616  white      F  hispanic   
4      16c73553-9400-1f2c-bcf9-7b5ef75150d2  95616  white      F  hispanic   
...                                     ...    ...    ...    ...       ...   
14841  30bfbb3e-2f6a-f599-c87e-3b3fdf0d0b33  90248  white      F  hispanic   
14842  30bfbb3e-2f6a-f599-c87e-3b3fdf0d0b33  90248  white      F  hispanic   
14843  30bfbb3e-2f6a-f599-c87e-3b3fdf0d0b33  90248  white      F  hispanic   
14844  30bfbb3e-2f6a-f599-c87e-3b3fdf0d0b33  90248  white      F  hispanic   
14845  30bfbb3e-2f6a-f599-c87e-3b3fdf0d0b33  90248  white      F  hispanic   

       INCOME  HEALTHCARE_EXPENSES  HEALTHCARE_COVERAGE  \
0       45504            286834.48            130435.94   
1       45504            286834.48            130435.94   
2       45504            286834.48            130435.94   
3       45504            286834.48            130435.94   
4       45504            286834.48            130435.94   
...       ...                  ...                  ...   
14841   88949            737028.50            851526.10   
14842   88949            737028.50            851526.10   
14843   88949            737028.50            851526.10   
14844   88949            737028.50            851526.10   
14845   88949            737028.50            851526.10   

                              CONDITION  \
0      Severe anxiety (panic) (finding)   
1      Severe anxiety (panic) (finding)   
2      Severe anxiety (panic) (finding)   
3      Severe anxiety (panic) (finding)   
4      Severe anxiety (panic) (finding)   
...                                 ...   
14841  Severe anxiety (panic) (finding)   
14842  Severe anxiety (panic) (finding)   
14843  Severe anxiety (panic) (finding)   
14844  Severe anxiety (panic) (finding)   
14845  Severe anxiety (panic) (finding)   

                                      OBSERVED_CONDITION VALUE  
0      Generalized anxiety disorder 7 item (GAD-7) to...   3.0  
1      Generalized anxiety disorder 7 item (GAD-7) to...   3.0  
2      Generalized anxiety disorder 7 item (GAD-7) to...  21.0  
3      Generalized anxiety disorder 7 item (GAD-7) to...   2.0  
4      Generalized anxiety disorder 7 item (GAD-7) to...   3.0  
...                                                  ...   ...  
14841  Generalized anxiety disorder 7 item (GAD-7) to...   1.0  
14842  Generalized anxiety disorder 7 item (GAD-7) to...   2.0  
14843  Generalized anxiety disorder 7 item (GAD-7) to...   3.0  
14844  Generalized anxiety disorder 7 item (GAD-7) to...   8.0  
14845  Generalized anxiety disorder 7 item (GAD-7) to...   0.0  

[14846 rows x 11 columns]

Data cleaning on Synthetic medical data

Drop rows with ZIP = 0

In [59]:
anxiety_df = anxiety_observations[anxiety_observations.ZIP != 0]
anxiety_df.shape

(14334, 11)

In [60]:
# Number of cities in California
anxiety_df.groupby('ZIP').size()

ZIP
90001    208
90002      3
90005      4
90006     31
90007      9
        ... 
95949      3
95963     21
95971     76
96003     13
96117     40
Length: 460, dtype: int64

In [61]:
# Number of California CITY (inside COUNTY) 
print(patients[patients['ZIP'] != 0].groupby('CITY').size())
# print(patients[patients['ZIP'] != 0].groupby('COUNTY').size())

CITY
Acton            2
Adelanto         7
Agoura Hills     2
Alameda         10
Alamo            1
                ..
Yountville       3
Yreka            2
Yuba City       16
Yucaipa          9
Yucca Valley     8
Length: 574, dtype: int64


In [62]:
anxiety_observations.columns

Index(['Id', 'ZIP', 'RACE', 'GENDER', 'ETHNICITY', 'INCOME',
       'HEALTHCARE_EXPENSES', 'HEALTHCARE_COVERAGE', 'CONDITION',
       'OBSERVED_CONDITION', 'VALUE'],
      dtype='object')

In the **anxiety_df dataset** of California state in US, there are 460 city/zip codes. In the **vic_zipcode dataset** of Victoria state in Australia, there are 749 suburbs/zip code.

In [63]:
# Paste your own full file path
aus_zipcode = pd.read_csv("/Users/cindy_homefolder/Desktop/y1s2/ENGR90033/choropleth map/synthea 2023/australian data/zipcodes_au.csv", index_col=None)
aus_zipcode.shape

(18513, 6)

In [64]:
aus_zipcode

USPS   ST                NAME  ZCTA5  \
0      Australian Capital Territory  ACT  Unincorporated ACT    200   
1      Australian Capital Territory  ACT  Unincorporated ACT    200   
2      Australian Capital Territory  ACT  Unincorporated ACT   2540   
3      Australian Capital Territory  ACT  Unincorporated ACT   2540   
4      Australian Capital Territory  ACT  Unincorporated ACT   2540   
...                             ...  ...                 ...    ...   
18508             Western Australia   WA            Gosnells   6989   
18509             Western Australia   WA            Gosnells   6990   
18510             Western Australia   WA            Armadale   6991   
18511             Western Australia   WA       Victoria Park   6992   
18512             Western Australia   WA            Armadale   6997   

             LAT         LON  
0     -35.277700  149.119000  
1     -35.277700  149.118900  
2     -34.863483  150.616440  
3     -34.863483  150.616440  
4     -35.129878  150.707521  
...          ...         ...  
18508 -32.039597  116.009575  
18509 -32.060407  116.008461  
18510 -32.097880  116.095955  
18511 -31.964503  115.920395  
18512 -32.097880  116.095955  

[18513 rows x 6 columns]

Generate Victoria dataset and replace ZIP column in anxiety_df to Victoria postcode due to the unsolved custom module issue of Synthea. 

In [65]:
vic_zipcode = aus_zipcode[aus_zipcode['USPS'] == 'Victoria']
vic_zipcode.shape


(3540, 6)

In [66]:
vic_zipcode

USPS   ST               NAME  ZCTA5        LAT         LON
13034  Victoria  VIC          Melbourne   3000 -37.817403  144.956776
13035  Victoria  VIC          Melbourne   3001 -37.817403  144.956776
13036  Victoria  VIC              Yarra   3002 -37.818517  144.982207
13037  Victoria  VIC          Melbourne   3003 -37.810871  144.949592
13038  Victoria  VIC       Port Phillip   3004 -37.844246  144.970161
...         ...  ...                ...    ...        ...         ...
16569  Victoria  VIC    South Gippsland   8438   0.000000    0.000000
16570  Victoria  VIC    South Gippsland   8511   0.000000    0.000000
16571  Victoria  VIC  Greater Dandenong   8785 -38.016114  145.208504
16572  Victoria  VIC  Greater Dandenong   8785 -38.016114  145.208504
16573  Victoria  VIC          Melbourne   9999 -37.817403  144.956776

[3540 rows x 6 columns]

The dataset got no suburb level, only the state level in zip code column, ZCTA5. Therefore, multiple zip codes belong to one suburb. 

In [67]:
vic_zipcode.groupby('NAME').size()

NAME
Alpine           48
Ararat           40
Ballarat        105
Banyule          25
Bass Coast       71
               ... 
Wodonga          15
Wyndham          27
Yarra            26
Yarra Ranges     29
Yarriambiack     25
Length: 80, dtype: int64

In [68]:
vic_zipcode.groupby('ZCTA5').size()

ZCTA5
3000    1
3001    1
3002    1
3003    1
3004    3
       ..
8205    1
8438    1
8511    1
8785    2
9999    1
Length: 749, dtype: int64

In [69]:
# vic_zipcode[vic_zipcode.ZCTA5 == 9999]
# vic_zipcode[vic_zipcode.NAME == 'Yarra']

Consider the data distribution in two datasets and their different dimension, I will take SEIFA data into consider too. 

In [70]:
demog = pd.read_csv("/Users/cindy_homefolder/Desktop/y1s2/ENGR90033/choropleth map/synthea 2023/australian data/demographics_au.csv")
demog.head(), demog.shape

(      ID  COUNTY               NAME           STNAME  POPESTIMATE2015  \
 0  10050   10050             Albury  New South Wales            56036   
 1  10180   10180  Armidale Regional  New South Wales            29332   
 2  10250   10250            Ballina  New South Wales            46172   
 3  10300   10300          Balranald  New South Wales             2207   
 4  10470   10470  Bathurst Regional  New South Wales            43653   
 
              CTYNAME  TOT_POP  TOT_MALE  TOT_FEMALE     WHITE  ...  35..50  \
 0             Albury    56036  0.487401    0.512599  0.522188  ...  0.0980   
 1  Armidale Regional    29332  0.485068    0.514932  0.522188  ...  0.1096   
 2            Ballina    46172  0.486529    0.513471  0.522188  ...  0.1108   
 3          Balranald     2207  0.496602    0.503398  0.522188  ...  0.1156   
 4  Bathurst Regional    43653  0.507594    0.492406  0.522188  ...  0.1192   
 
    50..75  75..100  100..150  150..200  200..999  LESS_THAN_HS  HS_DEGREE  \


In [71]:
demog[demog['STNAME']=='Victoria']

ID  COUNTY                NAME    STNAME  POPESTIMATE2015  \
129  20110   20110              Alpine  Victoria            13156   
130  20260   20260              Ararat  Victoria            11821   
131  20570   20570            Ballarat  Victoria           113482   
132  20660   20660             Banyule  Victoria           127376   
133  20740   20740          Bass Coast  Victoria            40641   
..     ...     ...                 ...       ...              ...   
204  27260   27260             Wyndham  Victoria           296193   
205  27350   27350               Yarra  Victoria            91543   
206  27450   27450        Yarra Ranges  Victoria           157419   
207  27630   27630        Yarriambiack  Victoria             6511   
208  29399   29399  Unincorporated Vic  Victoria              946   

                CTYNAME  TOT_POP  TOT_MALE  TOT_FEMALE     WHITE  ...  35..50  \
129              Alpine    13156  0.494831    0.505169  0.563199  ...  0.1108   
130              Ararat    11821  0.530412    0.469588  0.563199  ...  0.1016   
131            Ballarat   113482  0.483469    0.516531  0.563199  ...  0.1204   
132             Banyule   127376  0.488663    0.511337  0.563199  ...  0.1128   
133          Bass Coast    40641  0.488940    0.511060  0.563199  ...  0.1052   
..                  ...      ...       ...         ...       ...  ...     ...   
204             Wyndham   296193  0.506845    0.493155  0.563199  ...  0.1124   
205               Yarra    91543  0.490961    0.509039  0.563199  ...  0.1096   
206        Yarra Ranges   157419  0.497037    0.502963  0.563199  ...  0.1048   
207        Yarriambiack     6511  0.509599    0.490401  0.563199  ...  0.0788   
208  Unincorporated Vic      946  0.581395    0.418605  0.563199  ...  0.1068   

     50..75  75..100  100..150  150..200  200..999  LESS_THAN_HS  HS_DEGREE  \
129  0.1000   0.1000    0.0802    0.0604    0.0604      0.402393   0.160584   
130  0.1008   0.1008    0.0992    0.0976    0.0976      0.402393   0.160584   
131  0.0916   0.0916    0.0776    0.0636    0.0636      0.402393   0.160584   
132  0.0996   0.0996    0.0874    0.0752    0.0752      0.402393   0.160584   
133  0.1060   0.1060    0.0968    0.0876    0.0876      0.402393   0.160584   
..      ...      ...       ...       ...       ...           ...        ...   
204  0.1148   0.1148    0.0974    0.0800    0.0800      0.402393   0.160584   
205  0.1240   0.1240    0.1002    0.0764    0.0764      0.402393   0.160584   
206  0.1104   0.1104    0.0998    0.0892    0.0892      0.402393   0.160584   
207  0.0984   0.0984    0.1242    0.1500    0.1500      0.402393   0.160584   
208  0.1072   0.1072    0.0998    0.0924    0.0924      0.402393   0.160584   

     SOME_COLLEGE  BS_DEGREE  
129      0.109593    0.32743  
130      0.109593    0.32743  
131      0.109593    0.32743  
132      0.109593    0.32743  
133      0.109593    0.32743  
..            ...        ...  
204      0.109593    0.32743  
205      0.109593    0.32743  
206      0.109593    0.32743  
207      0.109593    0.32743  
208      0.109593    0.32743  

[80 rows x 47 columns]

Take city as a common column, filter the Victoria city that occurs in both datasets, victoria zipcode and anxiety df, then spread the postcode by ID column in anxiety datasets.

In [72]:
# vic_zipcode['NAME'].isin(demog['CTYNAME']).value_counts() #True : 3413 has repeated values
print(demog['CTYNAME'].isin(vic_zipcode['NAME']).value_counts())
print('demog dataframe has 77 Suburbs in VIC.')

False    469
True      77
Name: CTYNAME, dtype: int64
demog dataframe has 77 Suburbs in VIC.


In [73]:
print('vic_zipcode dataframe has', vic_zipcode['NAME'].nunique(), 'Suburbs in VIC.')

vic_zipcode dataframe has 80 Suburbs in VIC.


Filter the Suburb with zip code to form a new dataframe for later combination with anxiety_df.

In [74]:
unified_vic_zipcode = vic_zipcode[vic_zipcode['NAME'].isin(demog['CTYNAME'])]

In [75]:
unified_vic_zipcode.shape, unified_vic_zipcode.head()

((3413, 6),
            USPS   ST          NAME  ZCTA5        LAT         LON
 13034  Victoria  VIC     Melbourne   3000 -37.817403  144.956776
 13035  Victoria  VIC     Melbourne   3001 -37.817403  144.956776
 13036  Victoria  VIC         Yarra   3002 -37.818517  144.982207
 13037  Victoria  VIC     Melbourne   3003 -37.810871  144.949592
 13038  Victoria  VIC  Port Phillip   3004 -37.844246  144.970161)

In [ ]:
anxiety_df.groupby('ZIP')['Id'].nunique(), anxiety_df.ZIP.nunique()
# Califonia has 460 suburbs, so after replacing, it should keep 460 suburbs

In [77]:
print(len(anxiety_df.groupby('Id').size()), "patients")

664 patients


In [78]:
print(unified_vic_zipcode.ZCTA5.nunique(), "postcodes")
print(unified_vic_zipcode.NAME.nunique(), "suburbs")

722 postcodes
77 suburbs


**664 patients's postcode need to be replaced using 722 available postcodes among 77 suburbs in Victoria.**

In [ ]:
# Create a mapping between 'Id' and 'ZCTA5'
# id_to_zip = dict(zip(anxiety_df['Id'].unique(), unified_vic_zipcode['ZCTA5'].unique()))
id_to_zip = dict(zip(anxiety_df['ZIP'].unique(), unified_vic_zipcode['ZCTA5'].unique()))
# len(id_to_zip)
# Use the map function to replace 'ZIP' values based on 'Id'
anxiety_df['ZIP'] = anxiety_df['ZIP'].map(id_to_zip)


In [ ]:
# Create a mapping between 'ZIP' and 'NAME'
zip_to_name = dict(zip(unified_vic_zipcode['ZCTA5'], unified_vic_zipcode['NAME']))
# len(zip_to_name)
# Add a new column 'Suburb' based on 'ZIP' values 
anxiety_df['Suburb'] = anxiety_df['ZIP'].map(zip_to_name)

### After replacement, let's have a look at the new dataset.

In [ ]:
anxiety_df.shape, anxiety_df.head()

In [98]:
anxiety_df.Id.nunique(), anxiety_df.ZIP.nunique(), anxiety_df.Suburb.nunique()

(664, 460, 59)

In [99]:
# anxiety_df.groupby('Id')['ZIP'].unique() # 1 to 1 relationships
anxiety_df.groupby('ZIP')['Id'].nunique()

ZIP
3000    2
3001    2
3002    2
3003    2
3004    1
       ..
3633    1
3634    1
3635    1
3636    1
3637    1
Name: Id, Length: 460, dtype: int64

In [100]:
anxiety_df.groupby(['ZIP','Suburb']).size()

ZIP   Suburb            
3000  Melbourne              12
3001  Melbourne               7
3002  Yarra                  25
3003  Melbourne              16
3004  Port Phillip            7
                           ... 
3633  Greater Shepparton      3
3634  Greater Shepparton    102
3635  Moira                   6
3636  Moira                   6
3637  Moira                   3
Length: 460, dtype: int64

In [102]:
# Proof if new dataset has the same number of patients
anxiety_df.groupby(['Id', 'ZIP']).size()

Id                                    ZIP 
012aea97-b365-89a3-9a5e-f01882baf2d9  3164    50
0135a402-de4f-03f7-9b57-b21235269361  3402     5
0141371c-5848-c0d1-153b-29ec44de0561  3174     2
02852ee9-4629-e1f6-116f-fa52d77f2845  3265    15
028b01e2-02dc-e79e-54e3-856f9d3a8ed7  3254    18
                                              ..
fceacc48-c665-c4e2-ec22-e3d589336a4e  3103     9
fe6df80f-acff-34b4-3f60-f9f4f09e8a1c  3468     1
fe6ec497-bb8d-f614-5bcd-25324a5edd53  3279    56
feeb4853-6336-a868-130f-b733e4a17aca  3156     2
fffe51a2-5f76-d521-6079-bce7b606b42e  3011     4
Length: 664, dtype: int64

In [103]:
anxiety_df.to_csv("vic_anxiety.csv", encoding='utf-8', index=False)